tinygraph/tinygraph.hpp

```diff
@@ -0,0 +1,132 @@
+#ifndef TINYGRAPH_HPP
+#define TINYGRAPH_HPP
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include <tinygraph/tinygraph.h>
+
+
+namespace tg {
+
+  enum edges_are_sorted_s { edges_are_sorted };
+  enum edges_are_unsorted_s { edges_are_unsorted };
+
+  class graph {
+    public:
+
+      // construct from vectors, sorted
+      graph(
+          const std::vector<std::uint32_t>& sources,
+          const std::vector<std::uint32_t>& targets,
+          edges_are_sorted_s)
+        : handle{::tinygraph_construct_from_sorted_edges(
+            sources.data(), targets.data(), sources.size()), &::tinygraph_destruct} { }
+
+      // construct from vectors, unsorted
+      graph(
+          const std::vector<std::uint32_t>& sources,
+          const std::vector<std::uint32_t>& targets,
+          edges_are_unsorted_s)
+        : handle{::tinygraph_construct_from_unsorted_edges(
+            sources.data(), targets.data(), sources.size()), &::tinygraph_destruct} { }
+
+      // construct from ptr, sorted
+      graph(
+          const std::uint32_t * sources,
+          const std::uint32_t * targets,
+          std::uint32_t n,
+          edges_are_sorted_s)
+        : handle{::tinygraph_construct_from_sorted_edges(
+            sources, targets, n), &::tinygraph_destruct} { }
+
+      // construct from ptr, unsorted
+      graph(
+          const std::uint32_t * sources,
+          const std::uint32_t * targets,
+          std::uint32_t n,
+          edges_are_unsorted_s)
+        : handle{::tinygraph_construct_from_unsorted_edges(
+            sources, targets, n), &::tinygraph_destruct} { }
```

Comment:
The graph constructors here are very low-level and require the user to provide and own the two sequences of from/to nodes. This has two disadvantages:
1. We will have double the peak memory since at one point there will be the user sequences in memory as well as our graph storing offset and target arrays, and
2. It's not a nice API where the graph can be constructed incrementally or edges added incrementally

We could solve this partially by adding some sort of `graph_builder` or similar e.g. with `.add_edge(s, t)` functions and a `.build()` function returning a frozen `graph` type then.